stdlib/public/runtime/SymbolInfo.h

```diff
@@ -0,0 +1,73 @@
+//===--- ImageInspection.h - Image inspection routines ----------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// This file includes routines that extract metadata from executable and
+/// dynamic library image files generated by the Swift compiler. The concrete
+/// implementations vary greatly by platform.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_SYMBOLINFO_H
+#define SWIFT_RUNTIME_SYMBOLINFO_H
+
+#include "swift/Runtime/Config.h"
+
+#include <cstdint>
+
+#if defined(_WIN32)
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <Windows.h>
+#endif
+
+#if defined(_WIN32) && !defined(__CYGWIN__)
+#include <DbgHelp.h>
+#elif SWIFT_STDLIB_HAS_DLADDR
+#include <dlfcn.h>
+#endif
+
+namespace swift {
+struct SymbolInfo {
+private:
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  SYMBOL_INFO_PACKAGE _package;
+#elif SWIFT_STDLIB_HAS_DLADDR
+  Dl_info _info;
+#endif
+
+public:
+  SymbolInfo() {}
+
+  /// Get the file name of the image where the symbol was found.
+  const char *getFileName() const;
+
+  /// Get the base address of the image where the symbol was found.
+  const void *getBaseAddress() const;
+
+  /// Get the name of the symbol.
+  const char *getSymbolName() const;
+
+  /// Get the address of the symbol.
+  const void *getSymbolAddress() const;
+
+  static int lookup(const void *address, SymbolInfo *outInfo);
```

Comment:
I don't particularly understand the reasoning for this.  Why not just `friend int lookupSymbol(const void *, SymbolInfo *)` and avoid the member and move the body into the free function?

Reply:
I wanted to preserve the old signature for convenience. I could just as easily have the call sites call the static member function directly.

Reply:
Either is fine I think.  I do have a slight preference for the friend'ed free-standing function.

Reply:
I went a different route and kept `lookupSymbol()` for now with the signature callers have been using (to reduce the impact on call sites in this PR.) But I changed `SymbolInfo::lookup()` to return an `llvm::Optional<SymbolInfo>` instead of `int` + `outInfo`. In a future PR (when I make `getFilename()` work on Windows) I'll update the call sites to use the new static function and delete the old free function.